### PR TITLE
`startDevWorker` Milestone 2 - implement `LocalRuntimeController`

### PR DESCRIPTION
**Pull request by mrbbot**
**What this PR solves / how to test:**

Fixes [DEVX-856](https://jira.cfdata.org/browse/DEVX-856)

**Author has addressed the following:**

- Tests
  - [x] Included
  - [ ] Not necessary because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] Included
  - [x] Not necessary because: internal refactor
- Associated docs
  - [ ] Issue(s)/PR(s):
  - [x] Not necessary because: internal refactor

**Note for PR author:**

We want to celebrate and highlight awesome PR review! If you think this PR received a particularly high-caliber review, please assign it the label `highlight pr review` so future reviewers can take inspiration and learn from it.

### CODE REVIEW

File: packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts [hunk 1]
```diff
@@ -1,41 +1,369 @@
...
+				);
+			if (this.#mf === undefined) {
+				this.#mf = new Miniflare(options);
+			} else {
+				await this.#mf.setOptions(options);
+			}
+			// All asynchronous `Miniflare` methods will wait for all `setOptions()`
+			// calls to complete before resolving. To ensure we get the `url` and
+			// `inspectorUrl` for this set of `options`, we protect `#mf` with a mutex,
+			// so only update can happen at a time.
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts:280]:
```suggestion
			// so only one update can happen at a time.
```

File: packages/wrangler/src/api/startDevWorker/ProxyController.ts [hunk 2]
```diff
@@ -362,7 +367,7 @@ export class ProxyController extends EventEmitter {
 	}
 	onReloadComplete(data: ReloadCompleteEvent) {
 		this.latestConfig = data.config;
-		this.latestBundle = data.bundle;
+		this.latestBundle = data.bundle; // TODO(now): update `latestBundle` type
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/api/startDevWorker/ProxyController.ts:370]: - [ ] update `latestBundle` type and remove comment

File: packages/wrangler/src/api/startDevWorker/types.ts [hunk 3]
```diff
@@ -94,13 +117,27 @@ export interface StartDevWorkerOptions {
...
+
+export type Module<ModuleType extends ModuleRule["type"] = ModuleRule["type"]> =
+	File<string | Uint8Array> & {
+		/** Name of the module, used for module resolution, path may be undefined if this is a virtual module */
+		name: string;
+		/** How this module should be interpreted */
+		type: ModuleType;
+	};
+
+// TODO: revisit this type
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/api/startDevWorker/types.ts:139]: when?

Comment (penalosa) [reply packages/wrangler/src/api/startDevWorker/types.ts:139]: This is meant for later—I'll clarify

File: packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts [hunk 4]
```diff
@@ -0,0 +1,886 @@
...
+): Promise<ReloadCompleteEvent> {
+	const [event] = await events.once(controller, "reloadComplete");
+	return event;
+}
+
+function joinUrlParts(parts: UrlOriginParts | UrlOriginAndPathnameParts): URL {
+	const pathname = "pathname" in parts ? parts.pathname : "";
+	const spec = `${parts.protocol}//${parts.hostname}:${parts.port}${pathname}`;
+	return new URL(spec);
+}
```
Comment (RamIdeas) [inline packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts:56]: This already exists in utils – `urlFromParts`

File: packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts [hunk 5]
```diff
@@ -1,41 +1,369 @@
...
 import type {
 	BundleCompleteEvent,
 	BundleStartEvent,
 	PreviewTokenExpiredEvent,
 	ReloadCompleteEvent,
 	ReloadStartEvent,
 } from "./events";
+import type { File, ServiceFetch, StartDevWorkerOptions } from "./types";
+
+async function getBinaryFileContents(file: File<string | Uint8Array>) {
```
[the remark this answers is not in the document]
Comment (penalosa) [reply packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts:21]: I tend to keep helpers in one file (unexported) until they're needed elsewhere, and then when I need a helper like this I'll have a look around the codebase and refactor to a shared location if needed.

In this case though, `getBinaryFileContents()` is pretty generic, and future PRs based off this branch (the remote runtime controller) move it to `packages/wrangler/src/api/startDevWorker/utils.ts`.

File: packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts [hunk 6]
```diff
@@ -1,41 +1,369 @@
...
+		dispatch_namespaces: undefined,
+		mtls_certificates: undefined,
+		logfwdr: undefined,
+		unsafe: undefined,
+	};
+
+	const fetchers: Record<string, ServiceFetch> = {};
+
+	for (const [name, binding] of Object.entries(event.config.bindings ?? {})) {
+		binding.type;
```
[the remark this answers is not in the document]
Comment (penalosa) [reply packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts:79]: Typo, thanks for catching!

File: packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts [hunk 7]
```diff
@@ -1,41 +1,369 @@
...
+
+export const DEFAULT_WORKER_NAME = "worker";
+function getName(config: StartDevWorkerOptions) {
+	return config.name ?? DEFAULT_WORKER_NAME;
+}
+
+async function convertToConfigBundle(
+	event: BundleCompleteEvent
+): Promise<MF.ConfigBundle> {
+	const bindings: CfWorkerInit["bindings"] = {
```
[the remark this answers is not in the document]
Comment (penalosa) [reply packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts:51]: It does—that PR is based off this one (and it's targeted to this branch). The remote runtime controller refactors this helper to be shared.

File: packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts [hunk 8]
```diff
@@ -1,41 +1,369 @@
...
+		}
+	}
+	onBundleComplete(data: BundleCompleteEvent) {
+		const id = ++this.#currentBundleId;
+		this.emitReloadStartEvent({
+			type: "reloadStart",
+			config: data.config,
+			bundle: data.bundle,
+		});
+		void this.#mutex.runWith(() => this.#onBundleComplete(data, id));
```
[the remark this answers is not in the document]
Comment (penalosa) [reply packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts:345]: The void is for eslint—we have a linting rule set up to catch hanging promises, and so we need to explicitly mark any intentionally hanging promises with `void`

File: packages/wrangler/src/api/startDevWorker/ProxyController.ts (hunk 2 above)
Comment (penalosa) [reply packages/wrangler/src/api/startDevWorker/ProxyController.ts:370]: I've removed this comment—we might want to update this type in the future, but we can address that if needed. It's definitely not a `TODO(now)` urgency refactoring

File: packages/wrangler/src/api/startDevWorker/types.ts (hunk 3 above)
Comment (penalosa) [reply packages/wrangler/src/api/startDevWorker/types.ts:139]: I've removed this comment—we might want to update this type in the future, but we can address that if needed. It's definitely not a `TODO(now)` urgency refactoring

File: packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (penalosa) [reply packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts]: https://github.com/cloudflare/workers-sdk/pull/4450/commits/44c727e39c8560cc960f720529505a393ff2de53

File: packages/wrangler/src/api/startDevWorker/types.ts (hunk 3 above)
Comment (penalosa) [reply packages/wrangler/src/api/startDevWorker/types.ts]: https://github.com/cloudflare/workers-sdk/pull/4450/commits/44c727e39c8560cc960f720529505a393ff2de53

File: packages/wrangler/src/api/startDevWorker/ProxyController.ts (hunk 2 above)
Comment (penalosa) [reply packages/wrangler/src/api/startDevWorker/ProxyController.ts]: https://github.com/cloudflare/workers-sdk/pull/4450/commits/44c727e39c8560cc960f720529505a393ff2de53

File: packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts (hunk 4 above)
Comment (penalosa) [reply packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts]: https://github.com/cloudflare/workers-sdk/pull/4450/commits/5dc5115a871e00f821adbe5e0aba3ffd5ad0635d

File: packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (andyjessop) [reply packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts:51]: Ah, missed that. Thanks.

File: packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts [hunk 9]
```diff
@@ -1,41 +1,368 @@
...
+		} else if (binding.type === "logfwdr") {
+			bindings.logfwdr ??= { bindings: [] };
+			bindings.logfwdr.bindings.push({ ...binding, name: name });
+		} else if (binding.type.startsWith("unsafe-")) {
+			bindings.unsafe ??= {
+				bindings: [],
+				metadata: undefined,
+				capnp: undefined,
+			};
+			bindings.unsafe.bindings?.push({ ...binding, name: name });
```
Comment (RamIdeas) [inline packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts:150]: This is missing the `type` property transformation from `unsafe-x` to just `x`

**Review by CarmenPopoviciu — APPROVED** [review]

**Comment by CarmenPopoviciu** [comment]
great work @penalosa 🚀

**Review by andyjessop — APPROVED** [review]
LGTM!

File: packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts (hunk 9 above)
Comment (penalosa) [reply packages/wrangler/src/api/startDevWorker/LocalRuntimeController.ts]: https://github.com/cloudflare/workers-sdk/pull/4450/commits/d7c2bd7c7755036a74be539ddefabe42516c2384

**Review by RamIdeas — APPROVED** [review]
LGTM there's some leftover things to deal with unsafe bindings but we can do that in a follow-up

**Review by petebacondarwin — COMMENTED** [review]
A little late but here are some comments.

File: packages/wrangler/src/api/startDevWorker/RemoteRuntimeController.ts [hunk 10]
```diff
@@ -33,9 +32,9 @@ export class RemoteRuntimeController extends RuntimeController {
 	// *********************
 
 	emitReloadStartEvent(data: ReloadStartEvent) {
-		this.emit("reloadComplete", data);
```
[the remark this answers is not in the document]
Comment (RamIdeas) [reply packages/wrangler/src/api/startDevWorker/RemoteRuntimeController.ts:36]: It wasn't used previously

[the remark this answers is not in the document]
Comment (penalosa) [reply packages/wrangler/src/api/startDevWorker/RemoteRuntimeController.ts:36]: This wasn't in use previously (neither of the runtime controllers were hooked up)

**Pull request merged**